libs/partners/mongodb/langchain_mongodb/index.py

```diff
@@ -0,0 +1,132 @@
+import logging
+from typing import Any, Dict, List, Optional
+
+from pymongo.collection import Collection
+from pymongo.errors import CollectionInvalid, OperationFailure
+
+from langchain_mongodb.utils import _FailCode
+
+logger = logging.getLogger(__file__)
+
+
+def _create_index_definition(
+    dimensions: int,
+    path: str,
+    similarity: str,
+    filters: Optional[List[Dict[str, str]]],
+) -> Dict[str, Any]:
+    return {
+        "fields": [
+            {
+                "numDimensions": dimensions,
+                "path": path,
+                "similarity": similarity,
+                "type": "vector",
+            },
+            *(filters or []),
+        ]
+    }
+
+
+def create_vector_search_index(
+    collection: Collection,
+    index_name: str,
+    dimensions: int,
+    path: str,
+    similarity: str,
+    filters: List[Dict[str, str]],
+) -> None:
+    """Experimental Utility function to create a vector search index
+
+    Args:
+        collection (Collection): MongoDB Collection
+        index_name (str): Name of Index
+        dimensions (int): Number of dimensions in embedding
+        path (str): field with vector embedding
+        similarity (str): The similarity score used for the index
+        filters (List[Dict[str, str]]): additional filters for index definition.
+    """
+    db = collection.database
+    index_definition = {
+        "definition": _create_index_definition(
+            dimensions=dimensions, path=path, similarity=similarity, filters=filters
+        ),
+        "name": index_name,
+        "type": "vectorSearch",
+    }
+
+    try:
+        result = db.command(
+            {"createSearchIndexes": collection.name, "indexes": [index_definition]}
+        )
+        logger.info(result)
+    except CollectionInvalid:
```

Comment:
Would CollectionInvalid ever be raised here? I'm not seeing how that's possible.

Reply:
`CollectionInvalid` is the exception type raised when the index already exists. It seems a bit misleading given that the collection itself isn't the issue, just the pre-existence of an index within the collection with the same name.